rules/CodeQuality/Rector/Class_/GetRepositoryServiceLocatorToRepositoryServiceInjectionRector.php

```diff
@@ -0,0 +1,190 @@
+use Rector\Doctrine\NodeAnalyzer\RepositoryClassResolver;
+use Rector\Naming\Naming\PropertyNaming;
+use Rector\NodeManipulator\ClassDependencyManipulator;
+use Rector\PhpParser\Node\Value\ValueResolver;
+use Rector\PHPStan\ScopeFetcher;
+use Rector\PostRector\ValueObject\PropertyMetadata;
+use Rector\Rector\AbstractRector;
+use Rector\StaticTypeMapper\ValueObject\Type\FullyQualifiedObjectType;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+final class GetRepositoryServiceLocatorToRepositoryServiceInjectionRector extends AbstractRector
+{
+    /**
+     * @var string
+     */
+    private const GET_REPOSITORY_METHOD = 'getRepository';
+
+    public function __construct(
+        private readonly ValueResolver $valueResolver,
+        private readonly PropertyNaming $propertyNaming,
+        private readonly RepositoryClassResolver $repositoryClassResolver,
+        private readonly ClassDependencyManipulator $classDependencyManipulator,
+        private readonly ReflectionProvider $reflectionProvider
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Turns $this->entityManager->getRepository(...) on entity that suppors service repository, to constructor injection',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * @ORM\Entity(repositoryClass="SomeRepository")
+ */
+class SomeEntity
+{
+}
+
+use Doctrine\Bundle\MongoDBBundle\Repository\ServiceDocumentRepository;
+
+final class SomeRepository extends ServiceDocumentRepository
+{
+}
+
+final class SomeClass
+{
+    public function run()
+    {
+        return $this->getRepository(SomeEntity::class)->find(1);
+    }
+}
+
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    public function __construct(
+        private ServiceDocumentRepository $someEntityRepository
+    ) {
+    }
+
+    public function run()
+    {
+        return $this->someEntityRepository->find(1);
+    }
+}
+CODE_SAMPLE
+                ),
+
+            ]
+        );
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Class_
+    {
+        if ($this->shouldSkipClass($node)) {
+            return null;
+        }
+
+        $repositoryPropertyMetadatas = [];
+
+        $this->traverseNodesWithCallable($node, function (Node $node) use (
+            &$repositoryPropertyMetadatas
+        ): ?PropertyFetch {
+            if (! $node instanceof MethodCall) {
+                return null;
+            }
+
+            if (! $this->isName($node->name, self::GET_REPOSITORY_METHOD)) {
+                return null;
+            }
+
+            if (count($node->getArgs()) !== 1) {
```

Comment:
`$node->isFirstClassCallable()` need to be skipped to avoid assert() error

Reply:
Good point :+1: